TreeView with Search Bar: rewrite (#1014)
DevExpress/devextreme-demos: JSDemos/Demos/TreeView/TreeViewWithSearchBar/description.md

To add a search bar to the TreeView and customize the search functionality, do the following:

> - Set the [searchEnabled](/Documentation/ApiReference/UI_Components/dxList/Configuration/#searchEnabled) property to **true**.

Reviewer: This and a couple of other links lead to the List API reference.

> - Use the [searchMode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchMode) property to specify whether items should contain (default), start with, or match the search string. In this example, you can change the **searchMode** in the drop-down menu.

Reviewer 2: We've already told that the `searchMode` property we need to change the search behavior. Why do we need to repeat the property name from the demo's point of view? I would have written something more human-oriented for a change.

```suggestion
- Use the [searchMode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchMode) property to specify whether items should contain (default), start with, or match the search string. In this example, you can switch between search modes in the drop-down menu.
```

> - Configure the [searchExpr](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchExpr) property to specify custom search data fields. The default search field is [text](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/items/#text). 

Reviewer 2: Can I use more than one data field to search against simultaneously? Did you use the "custom" word only to contrapose this scenario to "the default search field" which is `text`?

I would have said that the TreeView uses the `text` property to determine the data field for search. Then, I would have said that you can use the `searchExpr` property to specify the fields against which I should allow users to search. That is only if I have understood the relationship between the mentioned properties correctly.

BTW, can I use the `text` property if I want to use a certain dataset field as a single search source?

> - Configure the [searchExpr](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchExpr) property to specify custom search data fields. The default search field is [text](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/items/#text). 

Reviewer 3: `custom search data fields` Why do you add 'data' here? 

```suggestion
- Configure the [searchExpr](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchExpr) property to specify custom search fields. The default search field is [text](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/items/#text). 
```

Author: To highlight that these fields belong to data objects.

> The TreeView uses the TextBox component as the search bar. To customize it, specify [TextBox properties](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/) in the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) object.

Reviewer 2: The TreeView control uses the TextBox component as a search bar (in general). Alternatively, you could have said:

> The TreeView uses the TextBox component as its search bar.

The thing is that the TreeView uses the TextBox component in a search bar role. This is not a very specific search bar that the TreeView control actually has but rather a general search bar concept. Please do keep this in mind.

```suggestion
The TreeView uses the TextBox component as a search bar. To customize it, specify [TextBox properties](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/) in the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) object.
```

> The TreeView uses the TextBox component as the search bar. To customize it, use the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) property and include specific properties from the [TextBox configuration](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/).

Reviewer: ```suggestion
The TreeView uses the TextBox component as the search bar. To customize it, specify [TextBox properties](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/) in the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) object.
```

> We do not recommend that you enable search if the TreeView uses [virtual mode](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeView/VirtualMode/) or [custom logic](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeView/LoadDataOnDemand/) to load data on demand. In these cases, the TreeView may not have the full dataset, and the search results include only loaded nodes.

Reviewer 2: You can percieve the active mode as a state of the TreeView control. It does not use the mode, it can either be in this mode or have this mode as active. The point is that we do not recommend customers to enable the search functionality if the TreeView loads data on demand for whatever reason.

> In these cases, the TreeView may not have the full dataset, and the search results include only loaded nodes.

I'd like to draw your attention towards this sentence because it suggests that the TreeView may or may not have the full dataset at its disposal.  However, this is not quite true, because the TreeView will not have the full dataset in 99,9% of cases.

One more thing, why don't we recommend them to enable the search functionality? Are the search results incorrect? No. They are correct. However, they only reflect a subset, a small portion of the search results we would have if we'd loaded the full dataset in advance.

I would have rewritten this paragraph and would have said that recommend you to use search functionality with caution when the TreeView control loads data on demand which is the case when virtual mode is active or you employ custom data load algorithms.

> Note that only loaded nodes appear in the search results. Do not enable search if the TreeView uses the [createChildren](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#createChildren) function or [virtual mode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#virtualModeEnabled) since the search will produce partial results.

Reviewer: ```suggestion
We do not recommend that you enable search if the TreeView [loads data on demand](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeView/LoadDataOnDemand/) or uses [virtual mode](https://js.devexpress.com/Demos/WidgetsGallery/Demo/TreeView/VirtualMode/). In these cases, the TreeView does not have the full dataset, and the search results include only loaded nodes.
```

To add a search bar to the TreeView, set the [searchEnabled](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEnabled) property to **true**.

### Configure search properties
* Search mode         

> To appear in the search results, a node's text should contain the search string. Use the [searchMode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchMode) property to specify whether the text should start with or be equal to the search string instead. In this demo, you can use the drop-down list to switch between these modes.

Reviewer 4: >To appear in the search results, a node's text should contain the search string.

I suggest inverting the point: 'The search returns nodes whose text contains the search string'. This seems a bit more straightforward.

>specify whether the text should start with or be equal to the search string instead

And what about the *"contains"* search mode?

Author: > And what about the "contains" search mode?

This sentence accounts for the "contains" mode: 'The search returns nodes whose text contains the search string'.

* Search fields             
The TreeView searches through the [text](/Documentation/ApiReference/UI_Components/dxTreeView/Node/#text) node fields. To specify different field(s), use the [searchExpr](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchExpr) property.

> Note that only loaded nodes appear in the search results. If the TreeView uses the [createChildren](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#createChildren) function or [virtual mode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#virtualModeEnabled) to load data on demand, you should not add the search bar to prevent partial results. 

Reviewer 4: What do you think about imperative mood here?

```suggestion
Note that only loaded nodes appear in the search results. Do not enable search if the TreeView uses the [createChildren](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#createChildren) function or [virtual mode](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#virtualModeEnabled) since the search will produce partial results.
```

### Configure the search bar

> The TextBox represents the base component for the search bar. To customize it, use the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) property and include specific properties from the [TextBox configuration](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/).

Reviewer 4: ```suggestion
TreeView uses the TextBox component as the search bar.  To customize it, use the [searchEditorOptions](/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#searchEditorOptions) property and include specific properties from the [TextBox configuration](/Documentation/ApiReference/UI_Components/dxTextBox/Configuration/).
```
